services/sharedchannel/msg.go

```diff
@@ -6,48 +6,104 @@ package sharedchannel
 import (
 	"encoding/json"
 
+	"github.com/mattermost/mattermost-server/v5/mlog"
 	"github.com/mattermost/mattermost-server/v5/model"
 )
 
 // msgCache caches the work of converting a change in the Posts table to a remote cluster message.
 // Maps Post id to syncMsg.
 type msgCache map[string]syncMsg
 
-// syncMsg represents a change in content (post add/edit/delete, reaction add/remove).
+// syncMsg represents a change in content (post add/edit/delete, reaction add/remove, users).
 // It is sent to remote clusters as the payload of a `RemoteClusterMsg`.
 type syncMsg struct {
+	ChannelId string            `json:"channel_id"`
+	PostId    string            `json:"post_id"`
 	Post      *model.Post       `json:"post"`
+	Users     []*model.User     `json:"users"`
 	Reactions []*model.Reaction `json:"reactions"`
 }
 
-// postsToMsg takes a slice of posts and converts to a `RemoteClusterMsg` which can be
-// sent to a remote cluster
-func (scs *Service) postsToMsg(posts []*model.Post, cache msgCache) (model.RemoteClusterMsg, error) {
+func (sm syncMsg) ToJSON() ([]byte, error) {
+	b, err := json.Marshal(sm)
+	if err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func (sm syncMsg) String() string {
+	json, err := sm.ToJSON()
+	if err != nil {
+		return ""
+	}
+	return string(json)
+}
 
+// postsToMsg takes a slice of posts and converts to a `RemoteClusterMsg` which can be
+// sent to a remote cluster.
+func (scs *Service) postsToMsg(posts []*model.Post, cache msgCache, rc *model.RemoteCluster, lastSyncAt int64) (model.RemoteClusterMsg, error) {
 	syncMessages := make([]syncMsg, 0, len(posts))
 
 	for _, p := range posts {
+		if p.IsSystemMessage() { // don't sync system messages
+			continue
+		}
+
 		if sm, ok := cache[p.Id]; ok {
 			syncMessages = append(syncMessages, sm)
 			continue
 		}
 
-		reactions, err := scs.server.GetStore().Reaction().GetForPost(p.Id, true)
+		// any reactions originating from the remote cluster are filtered out
```

Comment:
Can filtering things out on a remote cluster basis create problems with the cached message?

Reply:
Not sure I follow.  The filter ensures that when a remote sends a reaction we don't send it back.  The remote already has it so I don't see any cache issues.  I may be missing something.

Reply:
Mainly I was worried that we could be excluding data just for one remote and since that would get cached we would be using always the same message for a given post.

Reply:
Ah, I think you're on to something.  If the remote cluster from which a post was sent happens to be the first one getting sync'd then an incomplete sync message is cached.  Also, the cache can have issues as users are filtered out.  The caching is a premature optimization that I will remove until a reliable mechanism can be devised.  That cache is likely also  why I'm getting redundant updates occasionally - a remote caches a full sync message that then gets sent to a remote that only needs a partial update.

Reply:
Done via https://github.com/mattermost/mattermost-server/pull/16783/commits/b4c8d01f8adc1ac87bc789e6c033569c13fc2a9b